[FEATURE] Document site sets (#4378)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/SiteHandling/SiteSets.rst

..  include:: /Includes.rst.txt
..  index:: Site handling; Site sets
..  _site-sets:

=========
Site sets
=========

Site sets ship parts of the site configuration as composable pieces. They are

> intended to deliver settings, TypoScript, TSconfig and reference enabled content

Reviewer: Maybe cross-reference links to TypoScript and TSConfig?

Is the "content block" reference already available, or only for future TYPO3 versions?

Author: woops, thanks for catching this! I think it is a future reference as the content blocks are not yet merged to my knowledge

Author: Done

blocks for the scope of a site.

Extensions can provide multiple sets in order to ship presets for different
sites or subsets (think of frameworks) where selected features are exposed
as a subset (example: `typo3/seo-xml-sitemap`).

..  contents:: Table of content
    :local:

..  _site-sets-definition:

Site set definition
===================

A set is defined in an extension's sub folder in :path:`Configuration/Sets/`, for
example :file:`EXT:my_extension/Configuration/Sets/MySet/config.yaml`.

The folder name in :path:`Configuration/Sets/` is arbitrary, significant
is the `name` defined in :file:`config.yaml`. The `name` uses a `vendor/name`
scheme by convention, and *should* use the same vendor as the containing
extension. It may differ if needed for compatibility reasons (for example when
sets are moved to other extensions). If an extension provides exactly one set

> that should have the same `name` as defined in :file:`composer.json`.

Reviewer: ```suggestion
, it should have the same `name` as defined in :file:`composer.json`.
```

> The :file:`config.yaml` for a set that is composed of three subsets looks as

Reviewer: Are subsets the "dependencies"? This example only has two dependencies?

follows:

..  code-block:: yaml
    :caption: EXT:my_extension/Configuration/Sets/MySet/config.yaml

    name: my-vendor/my-set
    label: My Set

    # Load TypoScript, TSconfig and settings from dependencies
    dependencies:
      - some-namespace/slider
      - other-namespace/fancy-carousel

..  _site-sets-usage:

> Using a site set as dependency in a site

Reviewer: Do we want to establish "Site Set" as an uppercased special term? Or "Site set"?

Author: In my experience the positive effect of introducing such case rules is minimal. It is more important to link to this chapter when mentioning the term

========================================

Sets are applied to sites via `dependencies` array in site configuration:

..  code-block:: yaml
    :caption: config/sites/my-site/config.yaml

    base: 'http://example.com/'
    rootPageId: 1
    dependencies:
      - my-vendor/my-set

Site sets can also be added to a site via the backend module
:guilabel:`Site Management > Sites`.

..  _site-sets-settings-definition:

Settings definitions
====================

Sets can define settings definitions which contain more metadata than just a
value: They contain UI-relevant options like `label`, `description`, `category`
and `tags` and types like `int`, `bool`, `string`, `stringlist`, `text` or

> `color`. These definitions are placed in :file:`settings.definitions.yaml`

Reviewer: Enhance this paragraph with something like "This YAML file resembles what extensions can provide via :file:`ext_emconf.txt` as additional configuration"? I think that could help people understand what this is tryint to enhance/replace?

Author: I don't really get what you are trying to explain here. To my knowledge there was never a ext_emconf.txt and the ext_emconf.php contains completely different information. Feel free to make a follow up, though

    settings:
      foo.bar.baz:
        label: 'My example baz setting'
        description: 'Configure baz to be used in bar'
        type: int
        default: 5

..  _site-sets-settings:

Settings for subsets
====================

Settings for subsets (for example to configure settings in declared dependencies)
can be shipped via :file:`settings.yaml` when placed next to the set file
:file:`config.yaml`.

Note that default values for settings provided by the set do not need to be
defined here, as defaults are to be provided within
:file:`settings.definitions.yaml`.

Here is an example where the setting `styles.content.defaultHeaderType` as
provided by `typo3/fluid-styled-content` is configured via
:file:`settings.yaml`:

..  code-block:: yaml
    :caption: EXT:my_extension/Configuration/Sets/MySet/settings.yaml

    styles:
      content:
        defaultHeaderType: 1

This setting will be exposed as site setting whenever the set
`my-vendor/my-set` is applied as dependency to a site configuration.

..  _site-sets-typoscript:

TypoScript provider
===================

TypoScript dependencies can be included via set dependencies. This mechanism is

> much more effective than the previous static_file_include's or manual

Reviewer: Reference "static_file_include"?

:typoscript:`@import` statements.

> TypoScript dependencies via sets are  automatically ordered and

Reviewer: ```suggestion
TypoScript dependencies via sets are automatically ordered and
```

        ..  code-block:: bash

            typo3/sysext/core/bin/typo3 site:sets:list

..  _site-sets-example-site-package:

Example: Using a set within a site package
==========================================

You can see an example of using a sets within a site package in the extension
:t3ext:`site-package`,
`t3docs/site-package (Source on GitHub) <https://github.com/TYPO3-Documentation/TYPO3CMS-Tutorial-SitePackage-Code>`__.

The site package example extension has the following file structure:

..  directory-tree::
    :show-file-icons: true

    *   Configuration

        *   Sets

            *   SitePackage

                *   config.yaml
                *   constants.typoscript
                *   page.tsconfig
                *   settings.yaml
                *   setup.typoscript

            *   ...

    *   Resources

        *   ...

    *   composer.json
    *   ...

..  _site-sets-example-site-package-set:

Defining the site set with a fluid_styled_content dependency
------------------------------------------------------------

As our example site package only contains one site set the name of that set
is the same like the composer name of the site package.

> The site package depends on EXT:fluid_styled_content. Therefore the two

Reviewer: Do we have documentation of that extension to link to?

sets provided by that system extension are included as dependencies:

..  include:: _Sets/_site-package/_config.rst.txt

If you need additional dependencies, you can find all available sets with the
console command :ref:`bin/typo3 site:sets:list <site-sets-cli>`.

..  _site-sets-example-usage:

Using the site set as dependency of a site
------------------------------------------

After the example site package is installed, you can include the site set
in your site configuration:

..  literalinclude:: _Sets/_site-package/_site_config.yaml
    :language: yaml
    :caption: config/sites/<some_site>/config.yaml | typo3conf/sites/<some_site>/config.yaml

..  _site-sets-example-typoscript:

Loading TypoScript via the site package's set
---------------------------------------------

The example site package also loads its TypoScript by placing the files
:file:`constants.typoscript` and :file:`setup.typoscript` into the folder of the
site set. These TypoScript files use :typoscript:`@import` statements to import

> files from the extension`s directory :path:`Configuration/TypoScript`:

Reviewer: ```suggestion
files from the extension's directory :path:`Configuration/TypoScript`:
```

..  include:: _Sets/_site-package/_setup.rst.txt

> Dependant TypoScript is included by including dependant sets and not by

Reviewer: ```suggestion
Dependant TypoScript is (referenced|collected|bundled|generated) by including dependant sets and not by
```

Maybe we can somehow prevent "included by including" wording somehow?

> Dependant TypoScript is included by including dependant sets and not by

Author: ```suggestion
Dependant TypoScript is included by the dependant sets and not by
```

TypoScript imports.

..  _site-sets-example-settings:

Using the site set to override default settings
-----------------------------------------------

You can use the file

> :file:`EXT:site_package/Configuration/Sets/SitePackage/settings.yaml` to

Reviewer: Maybe this could mislead people to think that EXT:site_package is an extension they get from TYPO3. We could maybe phrase "You can use the file ... of your custom sitepackage extension [link to sitepackage docs] ..."?
